### ISSUE TITLE
PostrgeSQL schema setting for migrations table

### ISSUE THREAD

**Issue by maxtryk** [amacneil/dbmate#110]
Hi Adrian!

Is there a way to set the name of the postgres schema in which `schema_migrations` is created?

**Comment by amacneil** [comment]
Hi! Not currently, we default to using the `public` schema. 

What's your use case? E.g. are you storing independent dbmate-managed databases in different schemas? Or do you just choose to store everything in a schema with a non-default name? If you were going to customize the schema name for dbmate where would you expect to set it?

**Comment by maxtryk** [comment]
>Or do you just choose to store everything in a schema with a non-default name?

Correct.

>If you were going to customize the schema name for dbmate where would you expect to set it?

It feels like a command line option would do it - this stuff shouldn't change between environments...

Thank you for your reply and happy holidays!

**Comment by Rested** [comment]
Would you want it to create the schema (if it does not exist) or not?

**Comment by Videl** [comment]
Seems the schema is created if needed when using this:
``
DATABASE_URL="postgres://postgres:password@db:5432/db?sslmode=disable&search_path=myschema"
``

But the ``schema_migrations`` is still in public. Maybe it could follow the schema setting?

**Comment by enverbisevac** [comment]
This is very useful in multitenant scenarios for Semi-isolated approach.

**Comment by RabidFire** [comment]
Hi @amacneil - First off, love the vision of this library. Appreciate all the hard work! We currently use golang-migrate, but want to introduce non-Go services and were looking for something more agnostic. 😄 

Our team needs this feature to use this library, and I'd be happy to send in a PR if you could provide some guidance and verify these steps:

### Use case
Our team has a bunch of microservices, each talking to a different schema within the same database. Each microservice manages its own migrations table. This allows separation of concerns and running migrations independently.

### Initial Research
Did some digging and saw that `public.schema_migrations` is [hard-coded](https://github.com/amacneil/dbmate/blob/ed9e57a4ade7a83584be46de364e6bcff1a5f611/pkg/dbmate/postgres.go#L168). In particular, it's only hard-coded for Postgres and no other database.

My first thought was to remove `public.` from `public.schema_migrations` everywhere so that it would work automatically when someone passes in a `search_path` parameter within `DATABASE_URL` (like @Videl [commented](https://github.com/amacneil/dbmate/issues/110#issuecomment-665915427)). However, I see this was introduced by #49 because of the way `pg_dump` works. 

### Possible Solutions
**Option A:** While dumping the schema, we could either use `SELECT current_schema()` or find the table_schema like so: `SELECT table_schema FROM information_schema.tables WHERE table_type = 'BASE TABLE' AND table_name = 'schema_migrations';` - I think finding the table_schema might be good for backward compatibility, in case someone has already set up their schema_migrations in public but was using the `search_path` option.

**Option B:** We could do something similar to golang-migrate's [`x-migrations-table`](https://github.com/golang-migrate/migrate/tree/master/database/postgres) option that gets passed down via the `MigrationsTable` field in dbmate.DB struct (`db.go`), which in turn can be set via `DBMATE_MIGRATIONS_TABLE` (or `--migrations-table` CLI option). This could default to `public.schema_migrations`. Not sure how non-Postgres DB drivers would work though.

**Option C:** If `search_path` exists, split by `,` and pick the first one as the schema name for `schema_migrations` table. (This is what `current_schema()` is set to in Postgres).

Side Note: I'm also thinking that we should additionally use `search_path` while dumping the schema. Currently, dbmate dumps the entire database using pg_dump, and not just the ones in the `search_path`. Seems like we need to pass `--schema <schema>` option to pg_dump to make this work. It would be useful if using `search_path=schema1,public` automatically passes `--schema schema1 --schema public` to pg_dump to only dump those specific schemas. Could do this in the same PR or a separate PR if you'd like. Basically, boils down to "Better search_path support for Postgres" or something. 😄

Please let me know your thoughts.

**Comment by amacneil** [comment]
Hi @RabidFire - thanks for the investigation and write up!

I think a combination of your suggestions could be helpful, since each serves a slightly different purpose. Each of these could be separate PRs.

**Option A**: It would be a relatively simple change to adjust our `dump` logic to auto-detect the schema rather than hard-coding `public`. I think it might be simpler to grep the output from `pg_dump` rather than querying information_schema, just to make sure that we use the exact same fully qualified table name that `pg_dump` chose to use. Alternatively, you might need to do something like adding `AND table_schema = ANY (CURRENT_SCHEMAS(false))` to your sql query so that we only discover a migrations table in visible schemas.

**Option C**: I think this makes sense, if a `search_path` is specified in the URL then we should use the first schema specified as the schema we create `schema_migrations` table in (assuming the table is not already found in one of the visible schemas). Note that specifying `search_path` in the URL query string is a golang `lib/pq`-specific feature (e.g. it doesn't work if you pass that URL to `psql` directly). However, I still think this seems like the easiest way to support schemas.

**Side Note**: I agree this also makes sense, if a search path is specified then we should pass only the named schemas to `pg_dump`.

**Option B**: In addition to all the above, I think we could also implement the ability to change the name of the `schema_migrations` table. I can see this being potentially useful for people using other databases too.

Basically, I'm happy to accept PRs for any or all of the above features. I think as described they would actually all work pretty nicely together, adding more flexibility while still doing what most people expect by default. However, this is just my initial thinking on it so please feel free to make any other suggestions.

**Comment by RabidFire** [comment]
Thanks @amacneil -

It seems like Option A & my side note basically address the same thing. I believe you've already taken care of this in #166 

It also seems like you've taken care of Option C in #167 :) It's more configurable than what I had started writing in my fork (https://github.com/amacneil/dbmate/compare/master...dashxhq:use-search-path). Basically, I felt we could just remove `public` as Postgres automatically will look at the search_path anyway. But #167 makes it more configurable, and the work there could be used to configure the migration table name as well (Option B). Tiny note: I'm not sure if `schema` is a standard way to pass the schema name (I've only ever seen `search_path`) but since both are supported, it works for me!

Apologies for not being able to contribute. In any case, you seem to have done it in a much better way than I would have! :) If there was a donation link, I would have bought you some coffee!

**Comment by amacneil** [comment]
Good point, I was thinking maybe `schema` in the URL could be more database-agnostic, but I'll leave it with simply `search_path` for now.

Your branch looks pretty similar to how mine started out. Thanks for the feedback! I also have another PR coming shortly (https://github.com/amacneil/dbmate/pull/169) to automatically create the schema when necessary.

**Comment by amacneil** [comment]
Also note I extracted the ability to rename the `schema_migrations` table entirely into a separate issue: https://github.com/amacneil/dbmate/issues/168

**Comment by amacneil** [comment]
@RabidFire and others: I would love help testing this new behavior before I publish a release, please feel free to comment on this issue or one of the related PRs if you get a chance to test it.

**Comment by RabidFire** [comment]
@amacneil Just reporting back that we've been testing against the latest master and it works well for us. Thanks!

**Comment by amacneil** [comment]
Thanks!
